codalab/lib/interactive_session.py

```diff
@@ -0,0 +1,258 @@
+        if len(containers) == 0:
+            raise RuntimeError('Could not find interactive session container with name: %s' % name)
+        return containers[0]
+
+    def _get_container_name(self):
+        return 'interactive-session-%s' % self._session_uuid
+
+    def _construct_final_command(self):
+        try:
+            candidate_commands = self._get_bash_history()
+        except:
+            print(
+                'The history of bash commands could not be retrieved at path: %s'
+                % InteractiveSession._BASH_HISTORY_CONTAINER_PATH,
+                file=self._stderr,
+            )
+            return ''
+
+        # If a user passed in an initial command, prepend it to list of possible commands to choose from
+        if self._initial_command:
+            candidate_commands.insert(0, self._initial_command + '\n')
+
+        # Write out the commands to choose from and the instructions out to a file
+        path = os.path.join(self._bundle_path, 'edit_commands.txt')
+        with open(path, 'w') as f:
+            for command in candidate_commands:
+                f.write(command)
+            f.write(InteractiveSession._CHOOSE_COMMAND_INSTRUCTIONS)
+
+        # Open an editor to allow users to choose commands
+        os.system('{} {}'.format(find_default_editor(), path))
+
+        # Extract out the final commands minus the instructions
+        commands = []
+        for line in open(path).read().splitlines():
+            command = line.lstrip().rstrip()
+            if command and not command.startswith(InteractiveSession._INSTRUCTIONS_DELIMITER):
+                commands.append(command)
+
+        final_command = '\n'.join(commands)
+        if final_command:
+            print('\nFinal constructed command:\n%s' % final_command, file=self._stdout)
+        return final_command
+
+    def _get_bash_history(self):
+        # Copies out .bash_history from the container to bundle_path
+        path = os.path.join(self._bundle_path, '.bash_history')
+        f = open(path, 'wb')
+        stream, _ = self._container.get_archive(InteractiveSession._BASH_HISTORY_CONTAINER_PATH)
+        for chunk in stream:
+            f.write(chunk)
+        f.close()
+
+        # Extract out a list of commands from .bash_history
+        commands = []
+        with open(path) as f:
+            for i, line in enumerate(f):
+                command = (
+                    line.rstrip(InteractiveSession._NULL_BYTE)
+                    if i > 0
+                    else line.split(InteractiveSession._NULL_BYTE)[-1]
+                )
+                if command:
+                    commands.append(command)
+        return commands
+
+
+def main():
+    # Example usage of InteractiveSession
+    session = InteractiveSession(
+        args.docker_image, manager=CodaLabManager(), dependencies=[], bundle_locations={}
+    )
+    session.start()
+    session.cleanup()
+
+
+if __name__ == '__main__':
```

Comment:
Is this even used (calling this file directly)? We should probably document its usage somewhere or remove it.